File: CSE-344_System-Programming_2025/Final_Multi-threaded_Distributed_Chat_and_File_Server/python/client/chatclient.py
#!/usr/bin/env python3
# Chat Client implementation
# Multi-threaded, TCP-based chat and file-sharing system

import socket
import threading
import json
import sys
import os
import base64
import time
import colorama
from colorama import Fore, Style

# Initialize colorama for cross-platform colored terminal output
colorama.init()

# Configuration constants
MAX_USERNAME_LEN = 16
MAX_ROOM_NAME_LEN = 32
MAX_FILE_SIZE = 3 * 1024 * 1024  # 3MB
ALLOWED_FILE_TYPES = ['.txt', '.pdf', '.jpg', '.png']

# Global variables
client_socket = None
username = None
current_room = None
exit_flag = threading.Event()

def validate_username(username):
    """Validate username (max 16 chars, alphanumeric only)"""
    import re
    if not username or len(username) > MAX_USERNAME_LEN:
        return False
    return bool(re.match(r'^[a-zA-Z0-9]+$', username))

def validate_room_name(room_name):
    """Validate room name (max 32 chars, alphanumeric only)"""
    import re
    if not room_name or len(room_name) > MAX_ROOM_NAME_LEN:
        return False
    return bool(re.match(r'^[a-zA-Z0-9]+$', room_name))

def is_valid_file(filename, file_size):
    """Check if file is valid (type and size)"""
    _, ext = os.path.splitext(filename)
    if ext.lower() not in ALLOWED_FILE_TYPES:
        return False, "Invalid file type. Allowed types: " + ", ".join(ALLOWED_FILE_TYPES)
    if file_size > MAX_FILE_SIZE:
        return False, f"File too large. Maximum size: {MAX_FILE_SIZE/1024/1024}MB"
    return True, ""

def encode_file(file_path):
    """Read and encode a file in base64"""
    try:
        with open(file_path, 'rb') as file:
            file_data = file.read()
            file_size = len(file_data)
            
            # Check file validity
            valid, error_msg = is_valid_file(file_path, file_size)
            if not valid:
                return None, None, error_msg
            
            # Encode file data
            encoded_data = base64.b64encode(file_data).decode('utf-8')
            return encoded_data, file_size, None
    except FileNotFoundError:
        return None, None, "File not found."
    except Exception as e:
        return None, None, f"Error reading file: {str(e)}"

def send_message(message_dict):
    """Send a message to the server"""
    try:
        client_socket.send(json.dumps(message_dict).encode('utf-8'))
        return True
    except Exception as e:
        print(f"{Fore.RED}Error sending message: {str(e)}{Style.RESET_ALL}")
        return False

def receive_messages():
    """Receive and process messages from the server"""
    global current_room, username
    
    while not exit_flag.is_set():
        try:
            data = client_socket.recv(1024*1024*4).decode('utf-8')  # 4MB buffer for file transfers
            if not data:
                print(f"{Fore.RED}Connection to server lost{Style.RESET_ALL}")
                exit_flag.set()
                break
            
            message = json.loads(data)
            message_type = message.get("type")
            
            if message_type == "server":
                print(f"{Fore.GREEN}[Server]: {message['message']}{Style.RESET_ALL}")
                
                # Update room status if message indicates joining a room
                if "joined the room" in message['message']:
                    room_name = message['message'].split("'")[1]
                    current_room = room_name
            
            elif message_type == "error":
                print(f"{Fore.RED}[Server]: {message['message']}{Style.RESET_ALL}")
            
            elif message_type == "broadcast":
                sender = message["username"]
                msg = message["message"]
                room = message["room"]
                print(f"{Fore.CYAN}[{room}] {sender}: {msg}{Style.RESET_ALL}")
            
            elif message_type == "whisper":
                sender = message["from"]
                msg = message["message"]
                print(f"{Fore.MAGENTA}[Whisper from {sender}]: {msg}{Style.RESET_ALL}")
            
            elif message_type == "room_notification":
                print(f"{Fore.YELLOW}[Room]: {message['message']}{Style.RESET_ALL}")
            
            elif message_type == "file":
                sender = message["from"]
                filename = message["filename"]
                file_data = message["file_data"]
                
                # Decode file data
                try:
                    decoded_data = base64.b64decode(file_data)
                    
                    # Save file
                    if not os.path.exists("downloads"):
                        os.makedirs("downloads")
                    
                    # Handle filename collision
                    base_name, ext = os.path.splitext(filename)
                    save_path = os.path.join("downloads", filename)
                    counter = 1
                    
                    while os.path.exists(save_path):
                        new_filename = f"{base_name}_{counter}{ext}"
                        save_path = os.path.join("downloads", new_filename)
                        counter += 1
                    
                    with open(save_path, 'wb') as file:
                        file.write(decoded_data)
                    
                    print(f"{Fore.GREEN}[File received] {sender} sent you '{os.path.basename(save_path)}'. Saved to {save_path}{Style.RESET_ALL}")
                
                except Exception as e:
                    print(f"{Fore.RED}Error saving received file: {str(e)}{Style.RESET_ALL}")
            
            elif message_type == "server_shutdown":
                print(f"{Fore.RED}[Server]: {message['message']}{Style.RESET_ALL}")
                exit_flag.set()
                break
            
        except json.JSONDecodeError:
            print(f"{Fore.RED}Error: Received invalid data from server{Style.RESET_ALL}")
        except Exception as e:
            if not exit_flag.is_set():
                print(f"{Fore.RED}Error receiving messages: {str(e)}{Style.RESET_ALL}")
                exit_flag.set()
            break

def handle_command(command):
    """Handle user commands"""
    global current_room, username
    
    # Split the command into parts
    parts = command.split()
    
    if not parts:
        return
    
    cmd = parts[0].lower()
    
    # /join <room_name>
    if cmd == "/join":
        if len(parts) < 2:
            print(f"{Fore.RED}Usage: /join <room_name>{Style.RESET_ALL}")
            return
        
        room_name = parts[1]
        
        # Validate room name
        if not validate_room_name(room_name):
            print(f"{Fore.RED}Invalid room name. Must be alphanumeric and max 32 characters.{Style.RESET_ALL}")
            return
        
        # Send join command to server
        join_msg = {"type": "join", "room": room_name}
        send_message(join_msg)
    
    # /leave
    elif cmd == "/leave":
        if not current_room:
            print(f"{Fore.RED}You are not in any room.{Style.RESET_ALL}")
            return
        
        # Send leave command to server
        leave_msg = {"type": "leave"}
        send_message(leave_msg)
    
    # /broadcast <message>
    elif cmd == "/broadcast":
        if len(parts) < 2:
            print(f"{Fore.RED}Usage: /broadcast <message>{Style.RESET_ALL}")
            return
        
        if not current_room:
            print(f"{Fore.RED}You are not in any room.{Style.RESET_ALL}")
            return
        
        # Join all parts after command as message
        message = " ".join(parts[1:])
        
        # Send broadcast command to server
        broadcast_msg = {"type": "broadcast", "message": message}
        send_message(broadcast_msg)
    
    # /whisper <username> <message>
    elif cmd == "/whisper":
        if len(parts) < 3:
            print(f"{Fore.RED}Usage: /whisper <username> <message>{Style.RESET_ALL}")
            return
        
        target_user = parts[1]
        message = " ".join(parts[2:])
        
        # Send whisper command to server
        whisper_msg = {"type": "whisper", "target": target_user, "message": message}
        send_message(whisper_msg)
    
    # /sendfile <filename> <username>
    elif cmd == "/sendfile":
        if len(parts) != 3:
            print(f"{Fore.RED}Usage: /sendfile <filename> <username>{Style.RESET_ALL}")
            return
        
        filename = parts[1]
        target_user = parts[2]
        
        # Encode file
        file_data, file_size, error = encode_file(filename)
        if error:
            print(f"{Fore.RED}Error: {error}{Style.RESET_ALL}")
            return
        
        print(f"{Fore.YELLOW}Preparing to send file...{Style.RESET_ALL}")
        
        # Send file command to server
        file_msg = {
            "type": "sendfile",
            "filename": os.path.basename(filename),
            "target": target_user,
            "file_data": file_data,
            "file_size": file_size
        }
        send_message(file_msg)
    
    # /exit
    elif cmd == "/exit":
        print(f"{Fore.YELLOW}Disconnecting from server...{Style.RESET_ALL}")
        
        # Send exit command to server
        exit_msg = {"type": "exit"}
        send_message(exit_msg)
        
        # Set exit flag to stop receiving thread
        exit_flag.set()
    
    # Unknown command
    else:
        print(f"{Fore.RED}Unknown command. Type /help for available commands.{Style.RESET_ALL}")

def print_help():
    """Print help information"""
    print(f"{Fore.CYAN}===== Chat Client Help ====={Style.RESET_ALL}")
    print(f"{Fore.CYAN}/join <room_name>{Style.RESET_ALL} - Join or create a room")
    print(f"{Fore.CYAN}/leave{Style.RESET_ALL} - Leave the current room")
    print(f"{Fore.CYAN}/broadcast <message>{Style.RESET_ALL} - Send message to everyone in the room")
    print(f"{Fore.CYAN}/whisper <username> <message>{Style.RESET_ALL} - Send private message")
    print(f"{Fore.CYAN}/sendfile <filename> <username>{Style.RESET_ALL} - Send file to user")
    print(f"{Fore.CYAN}/exit{Style.RESET_ALL} - Disconnect from the server")
    print(f"{Fore.CYAN}/help{Style.RESET_ALL} - Show this help message")
    print(f"{Fore.CYAN}========================={Style.RESET_ALL}")

def main():
    """Main client function"""
    global client_socket, username, exit_flag
    
    # Check command line arguments
    if len(sys.argv) != 3:
        print("Usage: ./chatclient <server_ip> <port>")
        return
    
    server_ip = sys.argv[1]
    
    try:
        server_port = int(sys.argv[2])
    except ValueError:
        print("Port must be a number")
        return
    
    # Create client socket
    try:
        client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        client_socket.connect((server_ip, server_port))
    except Exception as e:
        print(f"Error connecting to server: {str(e)}")
        return
    
    print(f"{Fore.GREEN}Connected to server at {server_ip}:{server_port}{Style.RESET_ALL}")
    
    # Get username
    while True:
        username = input("Enter your username: ")
        if validate_username(username):
            break
        print(f"{Fore.RED}Invalid username. Must be alphanumeric and max 16 characters.{Style.RESET_ALL}")
    
    # Send login message
    login_msg = {"type": "login", "username": username}
    if not send_message(login_msg):
        return
    
    # Start thread to receive messages
    receive_thread = threading.Thread(target=receive_messages)
    receive_thread.daemon = True
    receive_thread.start()
    
    # Print welcome message and help
    print(f"{Fore.GREEN}Welcome to the chat client, {username}!{Style.RESET_ALL}")
    print_help()
    
    # Main input loop
    try:
        while not exit_flag.is_set():
            user_input = input("> ")
            
            if not user_input:
                continue
            
            if user_input.lower() == "/help":
                print_help()
                continue
            
            # Handle user command
            handle_command(user_input)
    
    except KeyboardInterrupt:
        print(f"{Fore.YELLOW}\nDisconnecting from server...{Style.RESET_ALL}")
        
        # Send exit command to server
        exit_msg = {"type": "exit"}
        try:
            send_message(exit_msg)
        except:
            pass
        
        exit_flag.set()
    
    except Exception as e:
        print(f"{Fore.RED}Error: {str(e)}{Style.RESET_ALL}")
    
    finally:
        # Clean up
        if client_socket:
            try:
                client_socket.close()
            except:
                pass
        
        print(f"{Fore.GREEN}Goodbye!{Style.RESET_ALL}")

if __name__ == "__main__":
    main()

File: CSE-344_System-Programming_2025/Final_Multi-threaded_Distributed_Chat_and_File_Server/c2/chatclient.c
/**
 * chatclient.c - Main client implementation
 * CSE 344 - System Programming
 */

#include "chatclient.h"

/* Global variables */
int client_socket = -1;
int client_running = 0;
ClientState client_state = STATE_DISCONNECTED;
char username[MAX_USERNAME_LENGTH + 1] = {0};
char current_room[MAX_MESSAGE_LENGTH] = {0};
pthread_mutex_t client_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_t receive_thread;
FileReceiveState file_state = {0};

/**
 * Signal handler
 */
void handle_signal(int sig) {
    if (sig == SIGINT) {
        printf("\nDisconnecting from server...\n");
        client_running = 0;
        
        /* Send exit command if connected */
        if (client_state >= STATE_CONNECTED) {
            send_to_server("/exit");
        }
        
        /* Close socket and clean up */
        disconnect_from_server();
        cleanup_client();
        
        exit(EXIT_SUCCESS);
    }
}

/**
 * Connect to the chat server
 */
int connect_to_server(const char* ip, int port) {
    struct sockaddr_in server_addr;
    
    /* Create socket */
    client_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (client_socket < 0) {
        perror("Socket creation failed");
        return -1;
    }
    
    /* Configure server address */
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);
    
    /* Convert IP address from text to binary form */
    if (inet_pton(AF_INET, ip, &server_addr.sin_addr) <= 0) {
        perror("Invalid IP address");
        close(client_socket);
        client_socket = -1;
        return -1;
    }
    
    /* Set client state to connecting */
    client_state = STATE_CONNECTING;
    
    /* Connect to server */
    if (connect(client_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        perror("Connection failed");
        close(client_socket);
        client_socket = -1;
        client_state = STATE_DISCONNECTED;
        return -1;
    }
    
    /* Create downloads directory if it doesn't exist */
    if (mkdir(DOWNLOAD_DIR, 0777) < 0 && errno != EEXIST) {
        perror("Failed to create downloads directory");
    }
    
    /* Set client state to connected */
    client_state = STATE_CONNECTED;
    client_running = 1;
    
    return 0;
}

/**
 * Disconnect from the server
 */
void disconnect_from_server(void) {
    pthread_mutex_lock(&client_mutex);
    
    if (client_socket >= 0) {
        close(client_socket);
        client_socket = -1;
    }
    
    client_state = STATE_DISCONNECTED;
    client_running = 0;
    
    pthread_mutex_unlock(&client_mutex);
}

/**
 * Clean up client resources
 */
void cleanup_client(void) {
    /* Close file if open */
    if (file_state.file) {
        fclose(file_state.file);
        file_state.file = NULL;
        file_state.active = 0;
    }
    
    /* Close socket if open */
    if (client_socket >= 0) {
        close(client_socket);
        client_socket = -1;
    }
    
    /* Destroy mutex */
    pthread_mutex_destroy(&client_mutex);
}

/**
 * Send message to server
 */
void send_to_server(const char* message) {
    pthread_mutex_lock(&client_mutex);
    
    if (client_socket >= 0 && client_state >= STATE_CONNECTED) {
        send(client_socket, message, strlen(message), 0);
    }
    
    pthread_mutex_unlock(&client_mutex);
}

/**
 * Thread function to receive messages from server
 */
void* receive_messages(void* arg __attribute__((unused))) {
    char buffer[CLIENT_BUFFER_SIZE];
    int bytes_read;
    
    while (client_running) {
        memset(buffer, 0, sizeof(buffer));
        
        pthread_mutex_lock(&client_mutex);
        if (client_socket < 0) {
            pthread_mutex_unlock(&client_mutex);
            break;
        }
        pthread_mutex_unlock(&client_mutex);
        
        /* Receive data from server */
        bytes_read = recv(client_socket, buffer, sizeof(buffer) - 1, 0);
        
        if (bytes_read <= 0) {
            /* Connection closed or error */
            if (client_running) {
                printf(COLOR_RED "\nDisconnected from server.\n" COLOR_RESET);
                disconnect_from_server();
            }
            break;
        }
        
        buffer[bytes_read] = '\0';
        
        /* Check if this is a file transfer command */
        if (strncmp(buffer, FILE_COMMAND_PREFIX, strlen(FILE_COMMAND_PREFIX)) == 0) {
            size_t header_size = 0;
            sscanf(buffer + strlen(FILE_COMMAND_PREFIX), "%zu", &header_size);
            handle_file_header(header_size);
            continue;
        }
        
        /* Check for server shutdown message */
        if (strcmp(buffer, "SERVER_SHUTDOWN") == 0) {
            printf(COLOR_RED "\nServer is shutting down. You will be disconnected.\n" COLOR_RESET);
            disconnect_from_server();
            break;
        }
        
        /* Regular message, print it */
        printf("%s\n", buffer);
        
        /* Check for login state transition */
        if (client_state == STATE_CONNECTED && 
            strstr(buffer, "Login successful") != NULL) {
            client_state = STATE_READY;
        }
    }
    
    return NULL;
}

/**
 * Handle file header received from server
 */
void handle_file_header(size_t header_size) {
    char sender[MAX_USERNAME_LENGTH + 1];
    char filename[MAX_MESSAGE_LENGTH];
    size_t file_size;
    char filepath[MAX_MESSAGE_LENGTH * 2];
    /* Buffer for messages - not used in current implementation */
    
    /* Receive file header */
    char* header_buffer = (char*)malloc(header_size);
    if (!header_buffer) {
        printf(COLOR_RED "Memory allocation error during file transfer.\n" COLOR_RESET);
        return;
    }
    
    ssize_t bytes_read = recv(client_socket, header_buffer, header_size, 0);
    if ((size_t)bytes_read != header_size) {
        printf(COLOR_RED "Error receiving file header.\n" COLOR_RESET);
        free(header_buffer);
        return;
    }
    
    /* Parse header information */
    memcpy(&file_size, header_buffer, sizeof(file_size));
    memcpy(filename, header_buffer + sizeof(file_size), MAX_MESSAGE_LENGTH);
    memcpy(sender, header_buffer + sizeof(file_size) + MAX_MESSAGE_LENGTH, MAX_USERNAME_LENGTH + 1);
    
    free(header_buffer);
    
    /* Check file size */
    if (file_size > MAX_FILE_SIZE) {
        printf(COLOR_RED "File too large: %zu bytes (max: %d).\n" COLOR_RESET, file_size, MAX_FILE_SIZE);
        return;
    }
    
    printf(COLOR_YELLOW "\nReceiving file '%s' from %s (%zu bytes)...\n" COLOR_RESET, 
           filename, sender, file_size);
           
    /* Check for duplicate filename and rename if necessary */
    snprintf(filepath, sizeof(filepath), "%s/%s", DOWNLOAD_DIR, filename);
    
    /* If file already exists, add a numeric suffix */
    int suffix = 1;
    char new_filepath[MAX_MESSAGE_LENGTH * 2];
    strcpy(new_filepath, filepath);
    
    while (access(new_filepath, F_OK) == 0) {
        /* File exists, try a new name */
        char* dot = strrchr(filename, '.');
        if (dot) {
            /* File has extension */
            char basename[MAX_MESSAGE_LENGTH];
            char extension[MAX_MESSAGE_LENGTH];
            
            /* Split into basename and extension */
            size_t basename_len = dot - filename;
            strncpy(basename, filename, basename_len);
            basename[basename_len] = '\0';
            strcpy(extension, dot);
            
            /* Create new filename with suffix */
            char basename_with_suffix[MAX_MESSAGE_LENGTH];
            snprintf(basename_with_suffix, sizeof(basename_with_suffix), "%s_%d", basename, suffix);
            snprintf(new_filepath, sizeof(new_filepath), "%s/%s%s", 
                     DOWNLOAD_DIR, basename_with_suffix, extension);
        } else {
            /* File has no extension */
            snprintf(new_filepath, sizeof(new_filepath), "%s/%s_%d", 
                     DOWNLOAD_DIR, filename, suffix);
        }
        
        suffix++;
    }
    
    /* Open file for writing */
    FILE* file = fopen(new_filepath, "wb");
    if (!file) {
        printf(COLOR_RED "Failed to open file for writing: %s\n" COLOR_RESET, new_filepath);
        return;
    }
    
    /* Set up file receiving state */
    file_state.active = 1;
    file_state.file = file;
    strncpy(file_state.filename, filename, MAX_MESSAGE_LENGTH - 1);
    strncpy(file_state.sender, sender, MAX_USERNAME_LENGTH);
    file_state.file_size = file_size;
    file_state.received_size = 0;
    
    /* Receive file data */
    char data_buffer[CLIENT_BUFFER_SIZE];
    size_t remaining = file_size;
    
    while (remaining > 0 && client_running) {
        size_t to_read = remaining > sizeof(data_buffer) ? sizeof(data_buffer) : remaining;
        bytes_read = recv(client_socket, data_buffer, to_read, 0);
        
        if (bytes_read <= 0) {
            printf(COLOR_RED "\nError receiving file data.\n" COLOR_RESET);
            break;
        }
        
        /* Write to file */
        if (fwrite(data_buffer, 1, (size_t)bytes_read, file) != (size_t)bytes_read) {
            printf(COLOR_RED "\nError writing to file.\n" COLOR_RESET);
            break;
        }
        
        remaining -= bytes_read;
        file_state.received_size += bytes_read;
        
        /* Show progress for large files */
        if (file_size > 1024 * 1024) {
            printf("\rReceiving: %zu/%zu bytes (%.1f%%)", 
                   file_state.received_size, file_size, 
                   (float)file_state.received_size / file_size * 100);
            fflush(stdout);
        }
    }
    
    /* Close file */
    fclose(file);
    file_state.file = NULL;
    file_state.active = 0;
    
    if (file_state.received_size == file_size) {
        /* Successful transfer */
        printf(COLOR_GREEN "\nFile received successfully: %s\n" COLOR_RESET, 
               strrchr(new_filepath, '/') ? strrchr(new_filepath, '/') + 1 : new_filepath);
    } else {
        /* Failed transfer */
        printf(COLOR_RED "\nFile transfer incomplete: %zu/%zu bytes received.\n" COLOR_RESET, 
               file_state.received_size, file_size);
        
        /* Remove incomplete file */
        unlink(new_filepath);
    }
}

/**
 * Process user input
 */
void process_user_input(void) {
    char input[MAX_MESSAGE_LENGTH];
    
    while (client_running) {
        /* Read user input */
        if (fgets(input, sizeof(input), stdin) == NULL) {
            break;
        }
        
        /* Remove trailing newline */
        size_t len = strlen(input);
        if (len > 0 && input[len - 1] == '\n') {
            input[len - 1] = '\0';
        }
        
        /* Skip empty input */
        if (strlen(input) == 0) {
            continue;
        }
        
        /* Process command or send message */
        if (client_state >= STATE_CONNECTED) {
            /* Send to server */
            send_to_server(input);
            
            /* Check for exit command */
            if (strcmp(input, "/exit") == 0) {
                client_running = 0;
                break;
            }
        }
    }
}

/**
 * Print welcome message
 */
void print_welcome_message(void) {
    printf(COLOR_CYAN);
    printf("╔══════════════════════════════════════════════╗\n");
    printf("║           Welcome to Chat Client             ║\n");
    printf("╚══════════════════════════════════════════════╝\n");
    printf(COLOR_RESET);
}

/**
 * Print help message
 */
void print_help(void) {
    printf(COLOR_YELLOW);
    printf("Available commands:\n");
    printf("  /join <room_name>       - Join or create a room\n");
    printf("  /leave                  - Leave the current room\n");
    printf("  /broadcast <message>    - Send message to everyone in the room\n");
    printf("  /whisper <user> <msg>   - Send private message\n");
    printf("  /sendfile <file> <user> - Send file to user\n");
    printf("  /exit                   - Disconnect from server\n");
    printf(COLOR_RESET);
}

/**
 * Main entry point
 */
int main(int argc, char* argv[]) {
    /* Check command line arguments */
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <server_ip> <port>\n", argv[0]);
        return EXIT_FAILURE;
    }
    
    /* Parse port number */
    int port = atoi(argv[2]);
    if (port <= 0 || port > 65535) {
        fprintf(stderr, "Invalid port number. Must be between 1-65535.\n");
        return EXIT_FAILURE;
    }
    
    /* Register signal handler */
    signal(SIGINT, handle_signal);
    
    /* Print welcome message */
    print_welcome_message();
    
    /* Connect to server */
    if (connect_to_server(argv[1], port) < 0) {
        fprintf(stderr, "Failed to connect to server at %s:%d\n", argv[1], port);
        return EXIT_FAILURE;
    }
    
    printf("Connected to server at %s:%d\n", argv[1], port);
    
    /* Start receive thread */
    if (pthread_create(&receive_thread, NULL, receive_messages, NULL) != 0) {
        perror("Failed to create receive thread");
        disconnect_from_server();
        return EXIT_FAILURE;
    }
    
    /* Process user input */
    process_user_input();
    
    /* Wait for receive thread to finish */
    if (receive_thread) {
        pthread_cancel(receive_thread);
        pthread_join(receive_thread, NULL);
    }
    
    /* Clean up */
    cleanup_client();
    
    printf("Disconnected from server.\n");
    return EXIT_SUCCESS;
}
